New setting to specify the start time of read plugins

rdietric opened the pull request:
Added the new setting _StartRead_, which specifies when the first read of each read plugin is performed. This allows the acquisition of metrics at specific times, e.g. at round points in time. It also allows staggered reading (even within a second) to prevent measurements from affecting each other.

The value is a float in seconds and interpreted as minutes, seconds and milliseconds of an hour, e.g. _StartRead 301.42_ performs the first read in minute 5, second 1 and millisecond 420 within the current hour. If the time has already passed, an initial read time in the future is determined by adding as many times _Interval_ as necessary.

The global value can be overwritten on a per-plugin base.

ChangeLog: New setting to specify the time of the first read for read plugins

rpv-tomsk commented on src/collectd.conf.in:
```diff
@@ -50,6 +50,17 @@
+#       StartRead 301.42 # minute 5, second 1, millisecond 420               #
+#   </LoadPlugin>                                                            #
+#----------------------------------------------------------------------------#
+#StartRead -1.0
```
1) This is not "StartRead" but "DelayRead"
2) Negative value has no sence.

rdietric replied:
1. Maybe "StartRead" was not the best choice. The value/option delays only the first read. So "DelayRead" may also be misleading. However, I do not insist on "StartRead"
2. Zero has a specific meaning: start reading at second 0 of each minute. This looks a lot nicer in timelines than starting at an arbitrary second.
3. Milliseconds enable explicit ordered execution of multiple plugins within a second, but avoid parallel reads which may affect each other. And we can still get the same epoche timestamp for all read values. (most reads are extremely fast, significantly below a second.

rpv-tomsk replied:
>Zero has a specific meaning: start reading at second 0 of each minute. 

1) I don't found this documented in provided PR
2) Typical Collectd setup calls read callbacks with 10 seconds interval. Once per minute is not enough frequency.
3) When there is not enough ReadThreads, this will not work.

>This looks a lot nicer in timelines than starting at an arbitrary second.

arbitrary second? Strange wish...

rpv-tomsk replied:
Milliseconds enable explicit ordered execution of multiple plugins within a second, but avoid parallel reads which may affect each other. And we can still get the same epoche timestamp for all read values. (most reads are extremely fast, significantly below a second.

From one side you want to serialize  read callbacks, from other side you want to get same epoch timestamp. contradiction.

>but avoid parallel reads which may affect each other.

Example, use-case?

rpv-tomsk commented on src/collectd.conf.pod:
```diff
@@ -139,6 +140,13 @@ defined in this plugin. By default, this is disabled.
+=item B<StartRead> I<Seconds>
+
+Sets a start time for the specific read plugin. The value is interpreted as 
+minutes, seconds and milliseconds of an hour. This plugin setting overrides the
```
1) "milliseconds of an hour" is strange.
2) milliseconds precision has no sence

`DelayRead - specifies plugin-specific delay before first metrics collection. This overrides the global DelayRead setting.`

Don't copy-n-paste text blocks you don't understand fully. There is no way for plugin to specify start time now and in your patch.

rpv-tomsk commented on src/daemon/plugin.h:
```diff
@@ -448,6 +449,17 @@ plugin_ctx_t plugin_set_ctx(plugin_ctx_t ctx);
+ *  return value will be a positive value in all cases. If
+ *  everything else fails, it will fall back to 0.
+ */
+double plugin_get_start_time(void);
```
I don't think this function has practical use-case.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,7 +1085,48 @@ static int plugin_insert_read(read_func_t *rf) {
+    /* if StartAt is not defined, start reading immediately */
+    rf->rf_next_read = cdtime();
+  } else {
+    /* get minute in hour and second+millisecond in minute */
```
IMHO: this code is overcomplicated. All you need is `rf->rf_next_read = cdtime() + rf->rf_ctx.first_read_delay; `

Default should be 0.

rdietric replied:
Do you mean the code that determines the `first_read_delay` is overcomplicated?
The if-branch takes care of the default case, when _StartRead_ has not been set. Thus, it executes `rf->rf_next_read = cdtime()` as done before. The else-branch determines when the next read should happen. This depends on the current time `cdtime()`, _StartRead_ and _Interval_ values.

rdietric replied:
> All you need is `rf->rf_next_read = cdtime() + rf->rf_ctx.first_read_delay; `

Determine the first_read_delay is the "complicated" part.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,7 +1085,48 @@ static int plugin_insert_read(read_func_t *rf) {
+    /* set the next read time */
+    if (sleep_cdt > 0) {
+      rf->rf_next_read = now + sleep_cdt;
+      INFO("%s plugin: wait %.3fms before first read at %ld.%03ds since epoche",
```
1) This will produce many similar lines when plugin registers several callbacks.
2) I don't think this message is useful for users, especially "123123123.123s since epoche" part.

IMHO: If you want, such message should be shown only once per plugin.

rdietric replied:
I made it a debug message. I did not find an easy way to show it only once per plugin.

rpv-tomsk commented:
>It also allows staggered reading (even within a second) to prevent measurements from affecting each other.

Internally, there is "read threads". So, all read calls are serialized.

If your measurements takes too long time (more than read interval), timeline may shift, and this makes overlap possible again.

What is your use-case about staggered readings / measurements affecting each other?

mrunge commented:
I have an issue with the name "StartRead". This is more about aligning plugin read. What about using AlignReadSeconds (or so?).

rdietric commented:
You are right, AlignRead is better. Should I split the single setting into AlignReadSeconds, AlignReadMinutes, AlignReadHours, AlignReadMilliseconds (AlignReadFractionsOfSecond). Currently, the value might be a little hard to interpret: 
> The StartRead float value is in seconds. Values smaller than 60 (and value mod 60) determine the second of a minute. Decimal places determine the millisecond of a second. Values greater or equal than 60 additionally determine the minute of an hour (the integer part of value divided by 60).

faxm0dem commented:
I am interested in this functionality for synchronizing reads across many machines. Will this ensure all reads are e.g. on the minute, or is a drift possible?

rdietric commented:
> Will this ensure all reads are e.g. on the minute, or is a drift possible?
This depends on the system times of the individual machines. It will drift with the drift of the system times between machines. In practice, if you use an NTP server on each machine, it should be at least accurate to seconds.

rpv-tomsk commented:
>>Will this ensure all reads are e.g. on the minute, or is a drift possible?
> In practice, if you use an NTP server on each machine, it should be at least accurate to seconds.

It can't be accurate to seconds in common case. If you have 10 read callbacks registered and 5 read threads, then 5 callbacks would wait until earlier-called callbacks do their work which can take some significant time.

rpv-tomsk commented:
AlignReadMilliseconds  option is stupid, you not on realtime OS. 
Based on this, I draw a negative conclusion about whole change.

rdietric commented:
> It can't be accurate to seconds in common case.

You are right. To get aligned and round timestamps, you have to consider the number of read threads, the execution time of read callbacks and the recording time of timestamps. Nevertheless, many reads callbacks use the proc filesystem and thus are executed in the microsecond range.

> AlignReadMilliseconds option is stupid, you not on realtime OS.
Based on this, I draw a negative conclusion about whole change.

This is not very productive. I see that you don't like the milliseconds with this new setting. I think that they have their use case and do not hurt. However, I could also live with removing the support of milliseconds and prioritizing the plugin read callbacks using a different setting.

I updated to initial comment to better reflect the intention of the new setting. I also changed the name of the setting to _AlignRead_ as suggested.

hahnjo commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1078,66 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+ * Values between 0 and 59 (mod 60) determine the second of a minute. Decimal
+ * places set the millisecond of a second.
+ */
+static void plugin_set_next_read_time(read_func_t *rf) {
```
Disclaimer: I'm not familiar with collectd's code. But from what I can see, `plugin_set_next_read_time` is only called once when the plugin is registered. To me, the current function name implies that it's called to determine _every_ `next_read_time`, so I'd propose to rename it to more clearly reflect the functionality - maybe `plugin_set_first_read_time`?

Slightly related, only taking `AlignRead` into account once might be a problem if system clock is reset. When synchronizing multiple systems over NTP, they all might have their own time drift and adjust differently at their own intervals. Not sure if that is something worth to consider...

rdietric replied:
I just renamed the function and replaced the "float" with "double". 

You might also be right with the time drift and/or clock reset. I am just thinking about how often doing the realignment. With another config setting, dependent on the AlignRead and/or Interval setting? As the realignment was on the critical path, I would not like to do it with every read.

rdietric replied:
Realignment of reads should happen automatically. Starting with the initial read time (absolute timestamp), which gets aligned with this feature, collectd determines the follow-up reads by adding the interval (or a multiple of the interval if a read took too long). Thus, all "next-reads" are absolute timestamps that are aligned.

faxm0dem commented:
On a side note, Sandia Lab developed [ldms](https://github.com/ovis-hpc/ovis) and one of the requirements was time synchronized metric reads across an hpc cluster. I've always thought they reinvented the wheel with respect to collectd, but they have timesync and RDMA

rdietric commented:
For us, the time synchronized/aligned metrics across compute nodes are also a requirement. What should I do to get this PR moving again?

mrunge commented on src/collectd.conf.pod:
```diff
@@ -9,6 +9,7 @@ collectd.conf - Configuration for the system statistics collection daemon B<coll
   BaseDir "/var/lib/collectd"
   PIDFile "/run/collectd.pid"
   Interval 10.0
+  StartRead -1.0
```
AlignRead?

rdietric replied:
Ups, I missed to rename this one. What about removing the line? This would have the same effect, unaligned reads as we currently have.

mrunge replied:
My preference would be to leave it there, with a comment on what it does. 

Having a parameter named AlignRead and a default value of -1.0 is not very obvious why and what the result would be.

rdietric replied:
I renamed it and added a comment, which also includes what 0.0 means. Furthermore, I added a sentence on the default value in the description of AlignRead.

rpv-tomsk commented:
>So I think that the motivation of synchronizing multiple instances of collectd is generally sound.

I do not understand people who want to add bursts/spikes of load to their systems.

Typical installations use average per interval, and there is no any practical difference between exact point of measure for them - requirement of exact interval between measurements is enough. I think this feature is not useful for the vast majority while it makes code complicated and harder to support.

rpv-tomsk commented:
>For us, the time synchronized/aligned metrics across compute nodes are also a requirement. What should I do to get this PR moving again?

Can you please tell us about your use-case with a detailed explanation of the reasons for the requirement? Why such requirement is critical for you?

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1078,66 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+
+    /* determine delay until next (aligned) read */
+    int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(start) - MS_TO_CDTIME_T(curr_msec);
+    while (sleep_cdt < 0) {
```
Why do you use loop here instead of mathematical expression?

rdietric replied:
I replaced the loop with an expression.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1078,66 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+    }
+
+    /* set the next read time */
+    if (sleep_cdt > 0) {
```
Unclear implementation.

Upper there is a loop, after that loop `sleep_cdt` could be zero or gt than zero (but not less).
So, both lines `rf->rf_next_read = now + sleep_cdt;` and `rf->rf_next_read = cdtime() + 0;` are nearby same, because of `now = cdtime()` upper. Why one line uses `now` and another uses `cdtime()` ?

rdietric replied:
The idea was to not have the initial rf->rf_next_read too much in the past, but for non-aligned reads, it will be in the past anyway. So you are right, the if/else is almost useless. I will remove it.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1078,66 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+ */
+static void plugin_set_first_read_time(read_func_t *rf) {
+
+  double start = rf->rf_ctx.align_read; /* (default is -1.0) */
```
Completely unclear naming: why `align_read` becomes `start` ?

rdietric replied:
I renamed the variable.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1078,66 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  } else {
+    cdtime_t now = cdtime();
+
+    /* get minute in hour and second in minute and millisecond in second */
```
Completely unclear comment.

It absolutely hard to understand what it means:  `get miliseconds since start of hour`. (And I'm still unclear if that is really this value).

rdietric replied:
I replaced the comment some time ago.

rdietric commented:
> I do not understand people who want to add bursts/spikes of load to their systems.
Doing all measurements at the same time can be beneficial, e.g. the cache is only polluted once and not continuously. 

> Typical installations use average per interval, and there is no any practical difference between exact point of measure for them - requirement of exact interval between measurements is enough. I think this feature is not useful for the vast majority while it makes code complicated and harder to support.
What time stamp do you use for the recorded values? We take the time, when the read plugin reads the metric. Currently, these timestamps are not round, which makes it "less beautiful" to visualize. If you also need to aggregate values across multiple machines and you only use intervals, the aggregates become more fuzzy. 

> I think this feature is not useful for the vast majority while it makes code complicated and harder to support.
At least faxm0dem was interested in this feature and and in terms of code changes it is very manageable (basically one function gathering an aligned start time).

Our use case are parallel programs across machines (compute nodes). In high-performance computing, MPI programs are common. Such programs are often SPMD programs, which means that all processes execute the same code in an ideal world at the same time. Thus, it is reasonable to also collect metrics at the same time to detect differences between the execution on different nodes.

rpv-tomsk commented:
>Doing all measurements at the same time can be beneficial, e.g. the cache is only polluted once and not continuously.

I'm not interested in "can be" arguments. "can be" is not equal to real arguments, which is missing, as I understand.

>Currently, these timestamps are not round, which makes it "less beautiful" to visualize. 
You can simply round timestamps on visualisation, that will produce the same result.

>If you also need to aggregate values across multiple machines and you only use intervals, the aggregates become more fuzzy.

I don't think somebody who uses Collectd really requires precise ("0.0001%") values. 

>At least faxm0dem was interested in this feature and and in terms of code changes it is very manageable (basically one function gathering an aligned start time).

This piece of Collectd code completely crosses your efforts to get aligned reads:

```
    /* Check, if `rf_next_read' is in the past. */
    if (rf->rf_next_read < now) {
      /* `rf_next_read' is in the past. Insert `now'
       * so this value doesn't trail off into the
       * past too much. */
      rf->rf_next_read = now;
    }
```

so I don't think you really get expected results all the time, and the fact you don't see that as a problem only confirms you don't really need this feature.

rpv-tomsk commented:
>Our use case are parallel programs across machines (compute nodes). In high-performance
> computing, MPI programs are common. Such programs are often SPMD programs, which means that 
>all processes execute the same code in an ideal world at the same time. 

But we are not in ideal world and used computers are not ideal too.

>Thus, it is reasonable to also collect metrics at the same time to detect differences between the execution on different nodes.

The only argue will be results compared from aligned and unaligned Collectd setup.
I think, you will get same average even your metrics will be unaligned.

rdietric commented:
> > If you also need to aggregate values across multiple machines and you only use intervals, the aggregates become more fuzzy.
> 
> I don't think somebody who uses Collectd really requires precise ("0.0001%") values.

This is not only about precision. How do you aggregate (temporally) unaligned metrics over multiple machines? Depending on the interval you use, you can get complete different aggregates. Only the average over all values will stay the same.

> This piece of Collectd code completely crosses out your efforts to get aligned reads:
> 
> ```
>     /* Check, if `rf_next_read' is in the past. */
>     if (rf->rf_next_read < now) {
>       /* `rf_next_read' is in the past. Insert `now'
>        * so this value doesn't trail off into the
>        * past too much. */
>       rf->rf_next_read = now;
>     }
> ```
> 
> so I don't think you really get expected results all the time, and the fact you don't see that as a problem only confirms you don't really need this feature.

It is only a problem, if reads take more time than the specified interval (which is extremely unlikely to happen in our setup). But you are right, if _AlignRead_ is set, a multiple of the interval should be added to `rf->rf_next_read` instead of assigning `now`.

rpv-tomsk commented:
>This is not only about precision. How do you aggregate (temporally) unaligned metrics over 
>multiple machines? Depending on the interval you use, you can get complete different aggregates. 

That is only about precision, as there is no problem to aggregate values with timestamp rounding on visualisation.

>if AlignRead is set, a multiple of the interval should be added to rf->rf_next_read instead of assigning now.

... and that opens some questions and implementation will become more complex.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -578,10 +578,16 @@ static void *plugin_read_thread(void __attribute__((unused)) * args) {
+        rf->rf_next_read = now;
+      } else {
+        /* ensure read alignment, if enabled */
+        while (rf->rf_next_read < now) {
```
You remove one loop and put another instead it?

Facepalm.

rdietric replied:
Yes, because the loop body is very likely only executed once. An adequate expression would introduce more overhead in most cases due to the necessary division.

rpv-tomsk replied:
Sounds reasonable..

rpv-tomsk replied:
Please add comment about "previous rf_next_read value is aligned". That is unclear here.

Another point, where alignment could be lost, is:

`rf->rf_effective_interval = max_read_interval;` + `rf->rf_next_read += rf->rf_effective_interval;`

rdietric replied:
I changed the comment and added a check for the _max_read_interval_ value.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
  * of a minute. The integer part of the value divided by 60 determines the
  * minute of an hour. Optionally, decimal places can be used to set the
- * fraction of a second (up to milliseconds precision).
+ * fraction of a second (up to microseconds precision).
```
Do you know something about schedulers precision (at OS level)?
Are you seriously thinking OS assigns microseconds quantums to tasks??

Want to disappoint you: quantum has milliseconds dimension.

rdietric replied:
This is only the internally used precision from the timeval structure. I can extend the comment to make clear that the alignment will very probable not be accurate to microseconds.

rpv-tomsk replied:
I think `(up to microseconds precision)` just needs to be removed.

rdietric replied:
Done.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
-            (int)(CDTIME_T_TO_TIMEVAL(rf->rf_next_read).tv_usec / 1000));
-    } else {
-      rf->rf_next_read = cdtime();
+  /* Seconds and minutes of Interval and AlignRead have to be perfect divisors
```
Wow, code becomes more complex ?? )))

I think, `plugin_set_first_read_time()` call is not good point to check configuration correctness.
I hope, some days `-T` option will be recovered....

rdietric replied:
As complex read callbacks can specify an interval and the recurrence of the read alignment depends on the interval, the checks cannot be performed earlier.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+
+  /* put current minutes, seconds and fractions of a second in a microseconds
+   * value (with stripped hours, days, etc.) */
+  uint64_t now_us =
```
Ouch.

All you need here is:

- calculate `now_us` as microseconds since hour start
- compare it with `alignRead`. if `alignRead > now_us`, then next time is:
- now_tv - now_us + alignRead

You must learn how `cdtime_t` works and use it only.

rdietric replied:
That's what the code does, if you remove the `if (sleep_cdt < 0)` block. The if-block considers the case `alignRead < now_us`.

rpv-tomsk replied:
`now - us_since_hour_start + align_since_hour_start` is MUCH more understandable than ` sleep_cdt +=  rf->rf_interval * ceil((double)(-sleep_cdt) / (double)rf->rf_interval);` and does not require floating point math. Much understandable also means less error prone.

You need to learn how `cdtime_t` works and use it.

rdietric replied:
The case you are talking about is `alignRead < now_us`, which does not require floating point math in my code. The other case is a little bit more complicated, as we want to find the shortest delay until the next aligned time (as written in the comment). But you are right, I can give `now_us` a more self-explanatory name.

rpv-tomsk replied:
```
  /* determine the minimum delay until next (aligned) read */
  int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(alignRead) - US_TO_CDTIME_T(now_us);
  if (sleep_cdt < 0) {
    sleep_cdt +=
        rf->rf_interval * ceil((double)(-sleep_cdt) / (double)rf->rf_interval);
  }

  /* set the next read time */
  rf->rf_next_read = now + sleep_cdt;
```

This piece of code uses floating point.
----
If `sleep_cdt >=0`:  

1) sleep_cdt = alignRead - now_us;
2) rf_next_read = now + sleep_cdt  = now - now_us + alignRead.

where 
`now_us` is usec since hour start 
`now - now_us` is 'startOfHour' point 
and 
`alignRead` is alignRead + interval * 0  (by the fact, for first reading in hour).

---

these expressions lead us to:

The `sleep_cdt < 0` is only when `now_us > alignRead` i.e. first aligned point of hour is in past.
We need to know next aligned point (numbered as X starting at 0). 
Then math is nearby same:

sleep_cdt = now - now_us + alignRead + interval * X 

```
if (now_us > alignRead) {
sleep_cdt = now - now_us + alignRead + interval_us * int(now_us / interval_us) 
}
```
this math not require floating point calculation and that's why is better.

Thinking further, the `if (now_us > alignRead) {}` condition is not correct too.
The correct one is to compare `now_us` with `interval_us`, because `alignRead` could not be >=  `interval_us`.

So we don't need this conditional here at all.

`sleep_cdt = now - now_us + alignRead + interval_us * int(now_us / interval_us)` fully covers all valid cases. Recall what `now - now_us` is `startOfHour`.

If `now_us < interval_us` then `int(now_us / interval_us)` will be 0 and first read will be at `startOfHour + alignRead * 1` which is expected.

If `alignRead == 0`  then all would be correct too.
Pure math. (cd_t and _us variables are mixed to simplify the appearance of the formulas).

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  cdtime_t now = cdtime();
+  struct timeval now_tv = CDTIME_T_TO_TIMEVAL(now);
+  struct tm *now_tm;
+  now_tm = localtime(&now_tv.tv_sec);
```
1) localtime() is not thread-safe.
2) look for  `CDTIME_T_TO_TIME_T`

rdietric replied:
The code is executed before read threads are started. Nevertheless, I thought about using get_local_time() from utils_time.c, which would require more code changes.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+
+  /* AlignRead values greater than Interval do not make sense. */
+  if (alignRead >= rf->rf_interval) {
+    static bool align_error = true;
```
Why you added this variable and condition check ????

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+      (sec_align > 0 && 60 % sec_align != 0) ||
+      (min_interval > 0 && 60 % min_interval != 0) ||
+      (min_align > 0 && 60 % min_align != 0)) {
+    static bool align_error = true;
```
>Why you added this variable and condition check ????

The same.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1085,48 +1091,78 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+
+    if (align_error) {
+      ERROR("AlignRead has to be smaller than Interval (%.3lf >= %d)!",
+           alignRead, (int)rf->rf_interval);
```
CDTIME_T_TO_DOUBLE(rf->rf_interval) should be used here.

rdietric replied:
Done.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  now_tm = localtime(&now_tv.tv_sec);
+
+  /* put current minutes, seconds and fractions of a second in a microseconds
+   * value (with stripped hours, days, etc.) */
```
A bothering comment again: "Calculate microseconds since start of current hour"
If you want: "Calculate microseconds since start of current hour (with stripped hours, days, etc.)"

You do not put minutes into this variable. This variable holds microseconds since start of current hour.  No minutes. No seconds.

rdietric replied:
I changed the comment.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  uint64_t now_us = (uint64_t)(now_tm->tm_min * 60 + now_tm->tm_sec) * 1000000 +
+                    now_tv.tv_usec;
+
+  /* determine the minimum delay until next (aligned) read */
```
Finally, taking into account existing piece of code above and your comment

>Yes, because the loop body is very likely only executed once. An adequate expression would introduce more overhead in most cases due to the necessary division.

I propose to recover your `while()` loop, to 
```
    /* determine delay until next (aligned) read */	
    int64_t sleep_cdt = DOUBLE_TO_CDTIME_T(alignRead) - US_TO_CDTIME_T(us_of_hour);	

    /* Calculated align point is in past, add several intervals to reach correct align point */
    while (sleep_cdt < 0) {	
      sleep_cdt += rf->rf_interval;	
    }
```

it would be better to have similar code.

rdietric replied:
I also think that the while loop is easier to understand. However, in the worst case we get 3600 iterations (e.g. AlignRead 0.1 and Interval 1). So I replaced the FP arithmetic in case of `alignRead < us_of_hour` with an integer version. If you still prefer the loop, I will recover it as in your proposal above.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+ * fraction of a second (up to microseconds precision).
+ */
+static void plugin_set_first_read_time(read_func_t *rf) {
+  double alignRead = rf->rf_ctx.align_read; /* (default is -1.0) */
```
Please rename `alignRead` to `align_read`, so naming style will be consistent.

rdietric replied:
Done.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  }
+
+  /* AlignRead values greater than Interval do not make sense. */
+  if (alignRead >= rf->rf_interval) {
```
rf_interval is cdtime_t. You could not compare them directly.

Please do validation and verify corner cases!

rdietric replied:
AlignRead is of type cdtime_t now (as you suggested).

rpv-tomsk commented on src/daemon/plugin.h:
```diff
@@ -173,6 +173,7 @@ typedef struct user_data_s user_data_t;
 
 struct plugin_ctx_s {
   char *name;
+  double align_read;
```
I'm still thinking if `align_read` should be of `cdtime_t` type....

rdietric replied:
Done.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+
+    if (align_error) {
+      ERROR("AlignRead has to be smaller than Interval (%.3lf >= %ld)!",
+            alignRead, CDTIME_T_TO_TIME_T(rf->rf_interval));
```
`CDTIME_T_TO_DOUBLE(rf->rf_interval)` should be used here!

rdietric replied:
Done.

rpv-tomsk commented on src/daemon/plugin.c:
```diff
@@ -1078,14 +1084,96 @@ static int plugin_compare_read_func(const void *arg0, const void *arg1) {
+  /* Seconds and minutes of Interval and AlignRead have to be perfect divisors
+   * of 60. Otherwise to perserve alignment. */
+  int sec_interval = CDTIME_T_TO_TIME_T(rf->rf_interval) % 60;
+  int sec_align = (long)alignRead % 60;
```
these /different/ casts looks strange.

rpv-tomsk commented:
Sorry for all the comments made.
I will no longer waste time reviewing this, simply because contributor does not review and does not work hardly on this feature.

Each review cycle reveals new and new problems that might not have been possible with due diligence. 

As the last example of insufficient elaboration, in addition to the comments already made, I will give an example of `AlignRead` value check.

Contributor states, what `AlignRead` value should be perfect divisor of 60. What is not true.
Absolutely no issues will occur with "AlignRead 7" (align to 7th second of minute) and interval, let's say, 10 seconds.


---
I think this PR is not good to be merged (nethier implementation, nor feature itself).
However, I have no merge rights (Developers are not welcomed here) and PR should be reviewed by somebody with such permissions.

Robert, thanks for all time you spent on my comments. Hope, you found it useful.

rdietric commented on src/daemon/plugin.c:
```diff
@@ -122,6 +122,11 @@ static char *plugindir;
 #define DEFAULT_MAX_READ_INTERVAL TIME_T_TO_CDTIME_T_STATIC(86400)
 #endif
+
+#ifndef DEFAULT_ALIGN_READ /* no alignment */
```
I changed it according to your suggestion.

rdietric commented on src/daemon/plugin.c:
```diff
@@ -1625,6 +1746,25 @@ EXPORT int plugin_init_all(void) {
+  if (align_read != DEFAULT_ALIGN_READ) {
+    /* make sure that MaxReadInterval does not brake read alignment */
+    if (!compliesReadAlign(max_read_interval)) {
+      WARNING("MaxReadInterval (%.3lf) might brake alignment. Fall back to "
```
I do not reset _MaxReadInterval_ any more and just print a warning. Correction to smaller value is reasonable, but the closest valid, smaller value is not easy to determine.

rdietric commented on src/daemon/plugin.c:
```diff
@@ -2545,6 +2685,15 @@ EXPORT cdtime_t plugin_get_interval(void) {
   return cf_get_default_interval();
 } /* cdtime_t plugin_get_interval */
 
+EXPORT double plugin_get_align_read(void) {
```
Sorry, there were still a few inconsistencies in the code, which I could have fixed before.

rdietric commented:
Thanks, Pavel, for the time you spent on reviewing my code. I made a full pass over the code and changed a couple of things, basically the checks. So I think your time was not wasted.

As there are a couple of different cases in the check for a recurring alignment, this introduces quite some lines of code. To be honest, I would leave it up to the user to set a reasonable interval, which does not destroy the alignment and remove the code that does the checks. This also makes the code changes of this PR easier to maintain.

rdietric commented:
As #1040 and #1197 propose a similar feature, there seems to be some interest. (Sorry for not looking carefully over the open PRs earlier.) Based on their solution and a minor extension, this PR allows more control over when read functions are called.

Background and reason (again) for this and also the other PRs is the monitoring of a compute cluster with thousands of nodes:
* same timestamps across different systems to avoid or simplify post-processing (might be performed multiple times)
* round timestamps (for [live] visualization of collected values)
* more control over the time when read functions are called (AlignReadOffset) - In our specific case we have a Likwid (https://github.com/RRZE-HPC/likwid) plugin, where the read duration is known and the timestamp is taken after the measurement (and should be a round timestamp).

faxm0dem commented:
Hi, this came up again at this year's HPCMASPA IEEE workshop, I would really like to see this merged, although I don't really have the skills to review it fully.
